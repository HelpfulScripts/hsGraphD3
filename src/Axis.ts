/**
 * # Axis class
 * 
 */

/** */
import { axisTop }          from 'd3';
import { axisRight }        from 'd3';
import { log as gLog }      from 'hsutil';   const log = gLog('Axis');
import { GraphComponent }   from './GraphComponent'; 
import { GraphCfg }         from './GraphComponent';
import { ComponentDefaults }from './GraphComponent';
import * as def             from './Defaults';
import { ScaleDefaults }    from './Scale';
import { UnitVp, d3Base }   from './Defaults';


export enum Direction {
    horizontal  = 'hor',
    vertical    = 'ver'
}

export interface AxisDefaults extends ComponentDefaults {
    color:      string;
    line:       def.Line;
    tickWidth:  UnitVp;
    tickMarks:  def.Line;
    tickLabel:  def.TextStyle;
}

export interface AxesDefaults extends ComponentDefaults {
    color:      string;
    hor:        AxisDefaults;
    ver:        AxisDefaults;
} 


export class Axes extends GraphComponent {
    private axes: Axis[] = [];

    constructor(cfg:GraphCfg) {
        super(cfg, cfg.baseSVG.select('.axes'));
        let axis;
        this.axes.push(new Axis(cfg, Direction.horizontal));
        this.axes.push(new Axis(cfg, Direction.vertical));
    }

    public get componentType() { return 'axes'; }

    public createDefaults():AxesDefaults {
        return {
            color:  '#000',
            hor:    this.axes[0].createDefaults(),
            ver:    this.axes[1].createDefaults()
        };
    }
    
    public renderComponent() {
        this.axes.forEach(axis => axis.renderComponent());
    }
}


export class Axis {
    private dir: Direction;
    private cfg: GraphCfg;
    private svg: d3Base;

    constructor(cfg:GraphCfg, dir:Direction) {
        this.cfg = cfg;
        this.svg = cfg.baseSVG.select('.axes').append('g');
        this.dir = dir;
        this.svg.classed(`${dir}Axis`, true);
    }

    public createDefaults() {
        return {
            color:      'currentColor',
            line:       def.defaultLine(2),
            tickWidth:  10,
            tickMarks:  def.defaultLine(2),
            tickLabel:  def.defaultText()
        };
    }
    
    public renderComponent() {
        const horScales = this.cfg.scales.hor;
        const verScales = this.cfg.scales.ver;
        const style = this.cfg.defaults.axes[this.dir];
        let axis;
        const margins = (<ScaleDefaults>this.cfg.defaults.scales).margin;
        this.cfg.baseSVG.select('.axes')
            .attr('color', this.cfg.defaults.axes.color);
        this.svg
            .attr('stroke', style.line.color)
            .attr('stroke-width', style.line.width)
            .attr('stroke-opacity', style.line.opacity);

        if (this.dir===Direction.horizontal) {
            axis = axisTop(horScales);
            const yCrossing = Math.max(margins.left, Math.min(verScales(0), this.cfg.viewPort.height-margins.right));
            this.svg.attr("transform", `translate(0, ${yCrossing})`);
        } else {
            axis = axisRight(verScales);
            const xCrossing = Math.max(margins.top, Math.min(horScales(0), this.cfg.viewPort.width-margins.bottom));
            this.svg.attr("transform", `translate(${xCrossing}, 0)`);
        }
        axis.tickSize(style.tickWidth);
        this.svg
            .attr('color', style.color)
            .style('font-family',  style.tickLabel.font.family)
            .style('font-size',   `${style.tickLabel.font.size}px`)
            .style('font-style',  style.tickLabel.font.style)
            .style('font-weight', style.tickLabel.font.weight);
        this.svg.call(axis);
        this.svg.selectAll('text')
            .transition().duration(1000)
            .attr('stroke', style.tickLabel.color || 'currentColor')
            .attr('stroke-width', 1);
        }
}