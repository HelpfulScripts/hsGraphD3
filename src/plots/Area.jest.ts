import * as hsGraphD3   from '../';
import * as d3          from 'd3';
import { DataSet }      from '../Graph';
import { AxesDefaults } from '../Axis';

let clientWidth = 300;
const root = window.document.createElement("div");
root.style.width = `${clientWidth}px`;
root.style.height = "300px";

d3.selection.prototype.transition   = function(){ return this; };
d3.selection.prototype.duration     = function(){ return this; };
d3.selection.prototype.ease         = function(){ return this; };

const data:DataSet = {
    colNames:['time', 'volume', 'costs'], 
    rows:[    [-1,    0.2,      0.3], 
              [0.2,   0.7,      0.2], 
              [0.4,   0.1,      0.3],
              [0.6,  -0.2,      0.1], 
              [0.8,   0.3,      0.5], 
              [1,     0.2,      0.4]
    ]
};

function createGraph(root:any) {
    const graph = new hsGraphD3.GraphCartesian(root);
    graph.isRendered = () => true;
    (<AxesDefaults>graph.defaults.axes).hor.numTicksMinor = 10;
    (<AxesDefaults>graph.defaults.axes).hor.numTicksMajor = 2;
    (<AxesDefaults>graph.defaults.axes).ver.numTicksMinor = 10;
    (<AxesDefaults>graph.defaults.axes).ver.numTicksMajor = 2;
    return graph;
}

describe('Area', () => {
    let graph:hsGraphD3.GraphCartesian;
    beforeAll(() => {
        graph = createGraph(root);
        graph.addSeries('area', {x:'time', y:'volume'});
        graph.render(data);
    });
    it(`should have 'area' registered`, () =>
        expect(graph.seriesTypes).toContain('area')
    );
    it('should plot area', () =>
        expect(root).toMatchSnapshot()
    );
    it('should support settings changes', () => {
        graph.defaults.axes.color = '#666';
        expect(graph.defaults.axes.color).toBe('#666');
    });
});
