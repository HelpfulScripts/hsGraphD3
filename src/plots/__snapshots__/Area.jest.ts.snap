// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Area should plot area 1`] = `
<div
  style="width: 300px; height: 300px;"
>
  <svg
    class="baseSVG"
    height="100%"
    preserveAspectRatio="xMinYMin meet"
    viewBox="0 0 1000 700"
    width="100%"
  >
    <g
      class="canvas"
    >
      <rect
        class="graphArea"
        fill="#fff"
        fill-opacity="1"
        height="700"
        rx="0"
        ry="0"
        stroke="#00c"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
        width="1000"
      />
    </g>
    <g
      class="grids"
    >
      <g
        class="hor-major-Grid"
        fill-opacity="0"
        stroke="#444"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="20"
          x2="980"
          y1="526"
          y2="526"
        />
        <line
          x1="20"
          x2="980"
          y1="164"
          y2="164"
        />
      </g>
      <g
        class="hor-minor-Grid"
        fill-opacity="0"
        stroke="#eee"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="20"
          x2="980"
          y1="670"
          y2="670"
        />
        <line
          x1="20"
          x2="980"
          y1="598"
          y2="598"
        />
        <line
          x1="20"
          x2="980"
          y1="526"
          y2="526"
        />
        <line
          x1="20"
          x2="980"
          y1="453"
          y2="453"
        />
        <line
          x1="20"
          x2="980"
          y1="381"
          y2="381"
        />
        <line
          x1="20"
          x2="980"
          y1="309"
          y2="309"
        />
        <line
          x1="20"
          x2="980"
          y1="237"
          y2="237"
        />
        <line
          x1="20"
          x2="980"
          y1="164"
          y2="164"
        />
        <line
          x1="20"
          x2="980"
          y1="92"
          y2="92"
        />
        <line
          x1="20"
          x2="980"
          y1="20"
          y2="20"
        />
      </g>
      <g
        class="ver-major-Grid"
        fill-opacity="0"
        stroke="#444"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="20"
          x2="20"
          y1="670"
          y2="20"
        />
        <line
          x1="500"
          x2="500"
          y1="670"
          y2="20"
        />
        <line
          x1="980"
          x2="980"
          y1="670"
          y2="20"
        />
      </g>
      <g
        class="ver-minor-Grid"
        fill-opacity="0"
        stroke="#eee"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="20"
          x2="20"
          y1="670"
          y2="20"
        />
        <line
          x1="116"
          x2="116"
          y1="670"
          y2="20"
        />
        <line
          x1="212"
          x2="212"
          y1="670"
          y2="20"
        />
        <line
          x1="308"
          x2="308"
          y1="670"
          y2="20"
        />
        <line
          x1="404"
          x2="404"
          y1="670"
          y2="20"
        />
        <line
          x1="500"
          x2="500"
          y1="670"
          y2="20"
        />
        <line
          x1="596"
          x2="596"
          y1="670"
          y2="20"
        />
        <line
          x1="692"
          x2="692"
          y1="670"
          y2="20"
        />
        <line
          x1="788"
          x2="788"
          y1="670"
          y2="20"
        />
        <line
          x1="884"
          x2="884"
          y1="670"
          y2="20"
        />
        <line
          x1="980"
          x2="980"
          y1="670"
          y2="20"
        />
      </g>
    </g>
    <g
      class="axes"
      color="currentColor"
    >
      <g
        class="horAxis"
        color="currentColor"
        fill="currentColor"
        fill-opacity="0"
        font-family="sans-serif"
        font-size="16px"
        font-style="normal"
        font-weight="100"
        stroke="currentColor"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
        text-anchor="middle"
        transform="translate(0, 526)"
      >
        <path
          class="domain"
          d="M20.5,5V0.5H980.5V5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(20.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            -1
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(116.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            -0.8
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(212.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            -0.6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(308.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            -0.4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(404.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            -0.2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(500.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(596.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            0.2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(692.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            0.4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(788.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            0.6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(884.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            0.8
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(980.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            1
          </text>
        </g>
      </g>
      <g
        class="verAxis"
        color="currentColor"
        fill="currentColor"
        fill-opacity="0"
        font-family="sans-serif"
        font-size="16px"
        font-style="normal"
        font-weight="100"
        stroke="currentColor"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
        text-anchor="end"
        transform="translate(500, 0)"
      >
        <path
          class="domain"
          d="M-5,670.5H0.5V20.5H-5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(0,670.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            -0.2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,598.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            -0.1
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,526.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,453.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            0.1
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,381.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            0.2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,309.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            0.3
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,237.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            0.4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,164.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            0.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,92.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            0.6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,20.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            0.7
          </text>
        </g>
      </g>
    </g>
    <g
      class="title"
    >
      <text />
    </g>
    <g
      class="series"
    >
      <g
        class="series0"
        style="color: rgb(27, 158, 119);"
      >
        <g
          class="area"
          fill="currentColor"
          fill-opacity="0.5"
        >
          <path
            d="M20,381C20,381,485.92675769484083,-4.356662656978037,596,20C697.0624876596856,42.36279106213441,657.7347758504364,341.7950439048414,692,453C722.23616686992,551.1289833266177,757.3251854673945,684.6483995906963,788,670C821.5889314028632,653.9599979165114,847.4592444493808,345.60125139642537,884,309C913.1168798285936,279.8349150851696,980,381,980,381L980,526C980,526,916,526,884,526C852.0000000000001,526,820,526,788,526C756,526,724.0000000000001,526,692,526C660,526,670.0304838915058,526,596,526C490.0646664919945,526.0000000000001,20,526,20,526"
          />
        </g>
      </g>
    </g>
  </svg>
</div>
`;
